### PR TITLE
[ARMv7][32bit] ASSERTION FAILED: CacheableIdentifier::isCacheableIdentifierCell(fieldNameValue)

### PR DESCRIPTION
**Pull request by mikhailramalho**
#### 22a4b34d6231dea23bdbf492e0c24e133f4c6e5d
<pre>
[ARMv7][32bit] ASSERTION FAILED: CacheableIdentifier::isCacheableIdentifierCell(fieldNameValue)
<a href="https://bugs.webkit.org/show_bug.cgi?id=242640">https://bugs.webkit.org/show_bug.cgi?id=242640</a>

Reviewed by Yusuke Suzuki.

This patch adds AccessType::GetPrivateNameById, the corresponding slow
path code, and changes compileGetPrivateNameById to use it when calling
cachedGetById.

The patch also includes some minor changes:
1. Adds default member initializer to StructureStubInfo
2. Rename argument names in JSValueRegs::withTwoAvailableRegs to be clearer
3. Fix typos in ICEVENT and where they are called

* JSTests/stress/get-private-name-cell.js: Added.
(__f_0):
(__v_0.0.__c_0.prototype.field.this.init.init):
(__v_0.0.__c_0.prototype.getField):
(__v_0.0.__c_0):
* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::reset):
(JSC::slowOperationFromUnlinkedStructureStubInfo):
(JSC::StructureStubInfo::initializeFromUnlinkedStructureStubInfo):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::StructureStubInfo):
(JSC::appropriateOptimizingGetByIdFunction):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/jit/GPRInfo.h:
(JSC::JSValueRegs::withTwoAvailableRegs):
(JSC::JSValueRegs::payloadOnly):
* Source/JavaScriptCore/jit/ICStats.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):

Canonical link: <a href="https://commits.webkit.org/254938@main">https://commits.webkit.org/254938@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e5727b1434edc7e0e3d3df4097a6c5a57c1da57b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90754 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35333 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100057 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33833 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83101 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/96442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96409 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77567 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/26763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/69794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/82282 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34921 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77268 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32727 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26649 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3443 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36499 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79856 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38423 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17490 "Passed tests") |

### CODE REVIEW

**Comment by mikhailramalho** [comment]
~~As I mentioned in the commit message, the functions in JITCall and JITPropertyAccess don't set the CellTag because I couldn't produce a test case that needs it to be set.~~

~~Maybe I should? It's mostly to be safe since we are passing a JSR and the tag is whatever was in the tag register.~~

Scratch that, I added them just in case but it's easy to remove if suggested.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 1]
```diff
@@ -1073,7 +1082,10 @@ void SpeculativeJIT::compileGetById(Node* node, AccessType accessType)
 
         base.use();
 
-        cachedGetById(node->origin.semantic, baseRegs, resultRegs, stubInfoGPR, scratchGPR, node->cacheableIdentifier(), JITCompiler::Jump(), NeedToSpill, accessType);
+        JSValueRegsTemporary property(this);
+        fillPropertyRegs(m_jit, node->cacheableIdentifier(), property);
+
+        cachedGetById(node->origin.semantic, baseRegs, property.regs(), resultRegs, stubInfoGPR, scratchGPR, node->cacheableIdentifier(), JITCompiler::Jump(), NeedToSpill, accessType);
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:1088]: These changes are regressing code generation for 64bit environment. Instead of passing JSValueReg, we should pass appropriate EncodedJSValue in cachedGetById. So, let's revert `DFGSpeculativeJIT.cpp` change.

**Review by Constellation — CHANGES_REQUESTED** [review]
I think we need different fix. The problem is not missing CellTag. The problem is that `AccessType::GetPrivateName` is used with cachedGetById while it is requesting JSValue for the property. (This is aligned with GetByVal).

The right fix is adding `AccessType::GetPrivateNameById`. And adding necessary changes to StructureStubInfo, AccessCase etc. And use it `AccessType::GetPrivateNameById`.

**Comment by mikhailramalho** [comment]
hey, @Constellation I addressed your comment and it's indeed a better solution.

The only thing I didn't like it the following code for 32-bit systems:
```
    JSValueRegsTemporary propertyTemp(this);
    JSValueRegs property = propertyTemp.regs();
    m_jit.move(SpeculativeJIT::TrustedImmPtr(identifier.rawBits()), property.payloadGPR());
    m_jit.move(identifier.isCell() ? MacroAssembler::TrustedImm32(JSValue::CellTag) : MacroAssembler::TrustedImm32(0), property.tagGPR());
```
If the identifier is not a cell, we could simply do `auto property = TrustedImmPtr(identifier.rawBits());` like how it's done for 64-bit systems, but I couldn't find a way of doing it without turning it into spaghetti code :/ Do you have any suggestions to fix that? 

Maybe a templated lambda could improve that, I'll try it tomorrow.

**Comment by mikhailramalho** [comment]
@Constellation the last version should do what you asked plus what I mentioned yesterday: we only generate the EncodedJSValue from two registers if the identifier is a cell, otherwise, we only use the identifier.rawbits().

**Comment by mikhailramalho** [comment]
ping @Constellation

**Review by Constellation — CHANGES_REQUESTED** [review]
As I said, the problem is not missing CellTag.
> I think we need different fix. The problem is not missing CellTag. The problem is that AccessType::GetPrivateName is used with cachedGetById while it is requesting JSValue for the property. (This is aligned with GetByVal).
> The right fix is adding AccessType::GetPrivateNameById. And adding necessary changes to StructureStubInfo, AccessCase etc. And use it AccessType::GetPrivateNameById.

Please follow to the above comment. I think we should not change DFGSpeculativeJIT.cpp much. The fix is adding AccessType::GetPrivateNameById and using it. And setting up IC code for that.

**Comment by mikhailramalho** [comment]
Hey @Constellation, I think I've addressed your comments, could you take another look?

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 2]
```diff
@@ -255,6 +255,79 @@ MacroAssembler::Jump SpeculativeJIT::emitOSRExitFuzzCheck()
     return result;
 }
 
+void SpeculativeJIT::cachedGetById(
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: Let's revert these changes. I think this is not necessary to this fix. We should make this patch simple.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 3]
```diff
@@ -4199,52 +4272,76 @@ void SpeculativeJIT::compileGetPrivateNameByVal(Node* node, JSValueRegs baseRegs
 
 void SpeculativeJIT::compileGetPrivateNameById(Node* node)
 {
-    switch (m_graph.child(node, 0).useKind()) {
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: Ditto, let's revert these changes. We should make this change small.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h [hunk 4]
```diff
@@ -718,15 +718,13 @@ class SpeculativeJIT {
 
     void compileCheckDetached(Node*);
 
-    void cachedGetById(CodeOrigin, JSValueRegs base, JSValueRegs result, GPRReg stubInfoGPR, GPRReg scratchGPR, CacheableIdentifier, JITCompiler::Jump slowPathTarget, SpillRegistersMode, AccessType);
+    void cachedGetById(CodeOrigin, JSValueRegs &base, JSValueRegs result, GPRReg stubInfoGPR, GPRReg scratchGPR, const CacheableIdentifier&, JITCompiler::Jump slowPathTarget, SpillRegistersMode, AccessType);
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h]: Ditto.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp [hunk 5]
```diff
@@ -168,67 +168,6 @@ bool SpeculativeJIT::fillJSValue(Edge edge, GPRReg& tagGPR, GPRReg& payloadGPR,
     }
 }
 
-void SpeculativeJIT::cachedGetById(CodeOrigin origin, JSValueRegs base, JSValueRegs result, GPRReg stubInfoGPR, GPRReg scratchGPR, CacheableIdentifier identifier, JITCompiler::Jump slowPathTarget , SpillRegistersMode mode, AccessType type)
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp]: Let's just use it.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp [hunk 6]
```diff
@@ -148,56 +148,6 @@ GPRReg SpeculativeJIT::fillJSValue(Edge edge)
     }
 }
 
-void SpeculativeJIT::cachedGetById(CodeOrigin origin, JSValueRegs base, JSValueRegs result, GPRReg stubInfoGPR, GPRReg scratchGPR, CacheableIdentifier identifier, JITCompiler::Jump slowPathTarget , SpillRegistersMode mode, AccessType type)
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp]: Let's just use it.

File: Source/JavaScriptCore/bytecode/StructureStubInfo.cpp [hunk 7]
```diff
@@ -536,6 +541,7 @@ void StructureStubInfo::initializeFromUnlinkedStructureStubInfo(const BaselineUn
         break;
     case AccessType::GetByVal:
     case AccessType::GetPrivateName:
+    case AccessType::GetPrivateNameById:
         hasConstantIdentifier = false;
```
Comment (Constellation) [inline Source/JavaScriptCore/bytecode/StructureStubInfo.cpp]: This is not correct since it has `hasConstantIdentifier`. Let's see how these values are used in `GetPrivateNameById` case and ensure that these configurations are correct.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: And I think FTL change is also required.

**Review by Constellation — CHANGES_REQUESTED** [review]
Since it is using JITGetByIdGenerator, this change needs to do appropriate things for appropriateGenericGetByIdFunction etc. And they should not list `GetPrivateName` then.
And let's make this patch smaller and simpler.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp (hunk 3 above)
Comment (mikhailramalho) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: I'm confused, do you want `compileGetPrivateNameById` to use `cachedGetById` again? But then we'll have the same problem of `appropriateOptimizingGetByIdFunction` expecting an `EncodedJSValue` but we are passing a `CacheableIdentifier`.

Comment (mikhailramalho) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: This is basically the same code from `compileGetPrivateNameByVal` but calling `operationGetPrivateNameByIdOptimize` instead of `operationGetPrivateNameOptimize`.

Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: > I'm confused, do you want compileGetPrivateNameById to use cachedGetById again?

Yes, it is what I'm suggesting so far. We should use `cachedGetById` with different AccessType.

>  But then we'll have the same problem of appropriateOptimizingGetByIdFunction expecting an EncodedJSValue but we are passing a CacheableIdentifier.

That's why I'm saying that you need to add a AccessType. You need to return the function which takes `CacheableIdentifier` from `appropriateOptimizingGetByIdFunction`.

Comment (mikhailramalho) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: Alright, updated.

**Comment by mikhailramalho** [comment]
@Constellation, indeed the solution you suggest is much smaller. Can you take another look? I think I finally addressed all your comments.

File: Source/JavaScriptCore/bytecode/StructureStubInfo.h [hunk 8]
```diff
@@ -473,6 +474,7 @@ inline auto appropriateOptimizingGetByIdFunction(AccessType type) -> decltype(&o
     case AccessType::GetByIdDirect:
         return operationGetByIdDirectOptimize;
     case AccessType::GetPrivateName:
```
Comment (Constellation) [inline Source/JavaScriptCore/bytecode/StructureStubInfo.h]: Why does `AccessType::GetPrivateName` happen?

File: Source/JavaScriptCore/bytecode/StructureStubInfo.h [hunk 9]
```diff
@@ -473,6 +474,7 @@ inline auto appropriateOptimizingGetByIdFunction(AccessType type) -> decltype(&o
...
+    case AccessType::GetPrivateNameById:
```
Comment (Constellation) [inline Source/JavaScriptCore/bytecode/StructureStubInfo.h:476]: Add handling of this in appropriateGenericGetByIdFunction too and drop `AccessType::GetPrivateName` from `appropriateGenericGetByIdFunction`.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 10]
```diff
@@ -4242,7 +4242,7 @@ void SpeculativeJIT::compileGetPrivateNameById(Node* node)
 
         JITCompiler::Jump notCell = m_jit.branchIfNotCell(baseRegs);
 
-        cachedGetById(node->origin.semantic, baseRegs, resultRegs, stubInfoGPR, scratchGPR, node->cacheableIdentifier(), notCell, NeedToSpill, AccessType::GetPrivateName);
+        cachedGetById(node->origin.semantic, baseRegs, resultRegs, stubInfoGPR, scratchGPR, node->cacheableIdentifier(), notCell, NeedToSpill, AccessType::GetPrivateNameById);
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:4235]: https://github.com/WebKit/WebKit/pull/3088#discussion_r974834669
> And I think FTL change is also required.

You need FTL change too.

**Review by Constellation — CHANGES_REQUESTED** [review]

File: Source/JavaScriptCore/bytecode/StructureStubInfo.h (hunk 8 above)
Comment (mikhailramalho) [reply Source/JavaScriptCore/bytecode/StructureStubInfo.h]: Removed.

File: Source/JavaScriptCore/bytecode/StructureStubInfo.h (hunk 9 above)
Comment (mikhailramalho) [reply Source/JavaScriptCore/bytecode/StructureStubInfo.h:476]: Added call to `operationGetPrivateNameByIdGeneric` in `appropriateGenericGetByIdFunction`.
Removed `AccessType::GetPrivateName` case

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp (hunk 10 above)
Comment (mikhailramalho) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:4235]: updated `compileGetPrivateNameById` to use `AccessType::GetPrivateNameById`.

**Review by Constellation — APPROVED** [review]
r=me
